Changes after v0.4

gshiroma opened the pull request:
This PR updates the DSWx-SAS in PROTEUS with changes after the v0.4:

Use basename for metadata fields: dem_file, landcover_file, and worldcover_file

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -2249,18 +2249,20 @@ def parse_runconfig_file(user_runconfig_file = None, args = None):
+        dem_file = os.path.basename(
+            ancillary_ds_group['dem_file'])
 
     if 'landcover_file' not in ancillary_ds_group:
         landcover_file = None
     else:
-        landcover_file = ancillary_ds_group['landcover_file']
+        landcover_file = os.path.basename(
+            ancillary_ds_group['landcover_file'])
 
     if 'worldcover_file' not in ancillary_ds_group:
         worldcover_file = None
     else:
-        worldcover_file = ancillary_ds_group[
-            'worldcover_file']
+        worldcover_file = os.path.basename(
+            ancillary_ds_group['worldcover_file'])
```
Unfortunately, changing the filepath here prevented the code from finding the source file for copying to the scratch directory. This is a fatal issue.

Instead, one idea could be to simply change the filenames that are written to the output  files. There might be a third place, but at minimum this would mean changing the code to  in two other places:
1) when writing the metadata dict
2) when writing entries into the log file

For example, to fix part 1, could maybe update Lines 2385-2392 to be something like this:

```
    # input datasets
    dswx_metadata_dict['HLS_DATASET'] = hls_dataset
    dswx_metadata_dict['DEM_FILE'] = os.path.basename(dem_file) if dem_file else '(not provided)'
    dswx_metadata_dict['LANDCOVER_FILE'] = \
        os.path.basename(landcover_file) if landcover_file else '(not provided)'
    dswx_metadata_dict['WORLDCOVER_FILE'] = \
        os.path.basename(worldcover_file) if worldcover_file \
                                     else '(not provided)'
```

gshiroma replied:
Thank you so much, @nemo794 ! This issue was addressed in [dbf9125](https://github.com/nasa/PROTEUS/pull/28/commits/dbf9125226a5baa9aad168c821e2ddd99e39e182).

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -878,6 +878,32 @@ def _apply_landcover_and_shadow_masks(interpreted_layer, nir,
         _is_landcover_class_evergreen(landcover_mask) &
         (nir > landcover_nir_threshold) & ((interpreted_layer == 3) |
                                            (interpreted_layer == 4)))
+    import plant
+
+    print('1:', landcover_mask[0:5, 0:5])
+    print('1:', landcover_mask.dtype)
+    print('2:', _is_landcover_class_evergreen(landcover_mask)[0:5, 0:5])
+    print('3:', landcover_mask[0:5, 0:5] == 201)
+    print('4:', (landcover_mask == 201)[0:5, 0:5])
+
+
+    '''
+
+    print( _is_landcover_class_evergreen(200))
+    print( _is_landcover_class_evergreen(201))
+    print( _is_landcover_class_evergreen(202))
+    plant.save_image(landcover_mask, output_file='cond_1_landcover_mask.tif', force=True)
+    print(_is_landcover_class_evergreen(landcover_mask))
+    print(np.nanmean(np.asarray(_is_landcover_class_evergreen(landcover_mask)), dtype=np.float32))
+
+    plant.save_image(np.asarray(_is_landcover_class_evergreen(landcover_mask), dtype=np.float32), output_file='cond_1_is_evergreen.tif', force=True)
+    plant.save_image(np.asarray(landcover_mask == 201, dtype=np.float32), output_file='cond_1_is_evergreen_2.tif', force=True)
+    plant.save_image(np.asarray(landcover_mask >= 201, dtype=np.float32), output_file='cond_1_is_evergreen_2_gt.tif', force=True)
+    plant.save_image(nir > landcover_nir_threshold, output_file='cond_2_nir.tif', force=True)
+    plant.save_image(interpreted_layer == 3, output_file='cond_3_intr_eq_3.tif', force=True)
+    plant.save_image(interpreted_layer == 4, output_file='cond_4_intr_eq_4.tif', force=True)
+    plant.save_image(interpreted_layer == 5, output_file='cond_5_intr.tif', force=True)
+    '''
```
I think this section is debug or experimental code? If so, should it be removed?

gshiroma replied:
Yes, debug code. Good catch! Thank you, @nemo794 !

nemo794 reviewed (changes requested):
Overall these are great improvements and fixes to the code! Nice! Thank you for adding them in.

As an aside, I noticed that this PR is failing to build in the CI system. Is this expected behavior? If not, is this something to look into?

gshiroma commented on bin/dswx_hls.py:
```diff
@@ -51,6 +51,7 @@ def main():
         args.output_file,
         hls_thresholds = hls_thresholds,
         dem_file=args.dem_file,
+        dem_description=args.dem_description,
```
Yes, good point. We need to decide what needs to be passed in this argument.

gshiroma replied:
I'm adding a comment to the DSWx-HLS product specs, and I'll resolve this comment here because it's not part of this PR, but thank you for the comment (we definitely need to define how the PGE will populate these fields).

nemo794 reviewed (commented):
Lots of good, hard work here - really nice! And thank you for addressing so many of the comments from before

The biggest issues I'm seeing have to do with maintainability --
First, while this is not the final delivery, let's remove dead code as we see it. For example, as written, the `_get_avg_sensing_time` function's logic is not doing what the name says it is, and this will be very confusing for future developers who will not know that.

Second, let's try to only have one place in a codebase to change a global value; the discussion in the comments about `min_slope_angle` are an example of this.

Third, regarding the `mask_adjacent_to_cloud_mode` feature, I thought that the decision had been made to not include that in this release because it is too experimental still? If so, let's remove that code entirely from this release to keep the main nasa/PROTEUS repo cleaner, and then add that code back to an e.g. changes_after_v0.5 branch.

Once again, lots of great work!

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -2981,7 +2981,7 @@ def generate_dswx_layers(input_list,
         if not os.path.isfile(filename):
             continue
-        # os.remove(filename)
+        os.remove(filename)
```
As far as I know, the `scratch_dir` or `scratch_path` is required for all SASes. I think it has to do with how the PGE maps the location of temporary files when executing the SAS within the Docker container.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -378,6 +469,15 @@ def get_dswx_hls_cli_parser():
                         help='Product ID that will be saved in the output'
                         "product's metadata")
 
+    parser.add_argument('--mask-adjacent-to-cloud-mode',
```
I think the decision is to get the opinion from the project (Steven C., etc) and decide what to do. We'll provide all three options, and we'll decide later what we should leave as default in the runconfig.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -648,63 +750,69 @@ def create_landcover_mask(copernicus_landcover_file,
+    # WorldCover class 95: mangroves
     water_binary_mask = np.where((worldcover_array_up_3 == 80) |
-                                 (worldcover_array_up_3 == 90), 1, 0)
-    h20_aggregate_sum = decimate_by_summation(water_binary_mask,
+                                 (worldcover_array_up_3 == 90) |
+                                 (worldcover_array_up_3 == 95), 1, 0)
```
I'm not sure if this is faster (or more efficient). Notice that if the condition matches one element, the `np.where()` will return a single index. The code that you are suggesting will return an array of `bool`.

nemo794 replied:
Before making the suggestion, I did a quick test with the existing code, and it is slightly faster. (20 iterations of this line took ~11 sec for the np.where() version, vs ~9 sec with the .astype() version.) Not a huge deal for an isolated instance like this, but since small things add up over the course of the codebase, still good to keep in mind.

Also, by using .astype(int), it should return the array of 1s and 0s as desired:

```
>>> x = np.array([[1,2,3],[4,3,2]])
>>> z = (x>2).astype(int)
>>> z
array([[0, 0, 1],
       [1, 1, 0]])
```

gshiroma replied:
Thank you for testing, @nemo794 . I'm still not convinced that `astype(int)` is  better. Benchmarks can be a bit tricky because of caching. Another thing that I forgot to mention is the memory usage. The condition should match a low percentage of the input pixels (as we are looking for certain landcover classes). So, the size of the element returned by `np.where()` should be much smaller than the one returned by `astype(int)` , which BTW probably uses much more memory than `bool` (that I initially had in mind). The resulting array is passed to another function, so I'm not sure if using `astype(int)` is really more efficient?

nemo794 replied:
I just tested again, and np.where() is actually returning an int64 array as well:

Debug Code:
```
    import time
    start = time.time()
    for i in range(20):
        water_binary_mask = np.where((worldcover_array_up_3 == 80) |
                                    (worldcover_array_up_3 == 90) |
                                    (worldcover_array_up_3 == 95), 1, 0)
    print('water_binary_mask time (sec): ', time.time() - start)
    print('water_binary_mask.dtype: ', water_binary_mask.dtype)
    print("Memory size of one array element of water_binary_mask in bytes: ", water_binary_mask.itemsize)
    print("Memory size of numpy array water_binary_mask in bytes:", water_binary_mask.size * water_binary_mask.itemsize)
 
    start = time.time()
    for i in range(20):
        water_binary_mask2 = ((worldcover_array_up_3 == 80) |
                                    (worldcover_array_up_3 == 90) |
                                    (worldcover_array_up_3 == 95)).astype(int)
    print('water_binary_mask2 time (sec): ', time.time() - start)
    print('water_binary_mask2.dtype: ', water_binary_mask2.dtype)
    print("Memory size of one array element of water_binary_mask2 in bytes: ", water_binary_mask2.itemsize)
    print("Memory size of numpy array water_binary_mask2 in bytes:", water_binary_mask2.size * water_binary_mask2.itemsize)
```

Output:
```
water_binary_mask time (sec):  21.7958083152771
water_binary_mask.dtype:  int64
Memory size of one array element of water_binary_mask in bytes:  8
Memory size of numpy array water_binary_mask in bytes: 964483200

water_binary_mask2 time (sec):  16.826853275299072
water_binary_mask2.dtype:  int64
Memory size of one array element of water_binary_mask2 in bytes:  8
Memory size of numpy array water_binary_mask2 in bytes: 964483200
```

Is there a different/better way to test this? Also, I'm curious why passing it to another array would make a difference in efficiency? (Does astype() do something funny under the hood?)

gshiroma replied:
Uhm, that's right, the code is not doing what I expected. I just did a quick test here and I noticed the following:

```
>>> np.where([1, 0, 0])
(array([0]),)
>>> np.where([1, 0, 0], 1, 0)
array([1, 0, 0])
```

Notice that the first command returns only the indices that match the condition and the second returns the full array. This happens because of the extra arguments `, 1, 0)`. I copied that part of the code from John's script and I didn't pay attention to it. I also noticed now that the code inside the called function requires an array, rather than list of indices. So yes, it makes sense to change the masking. I'll update the code soon. Thank you.

gshiroma replied:
Thank you again for finding this opportunity to improve the code. I just substituted `astype(int)`  with `astype(np.uint8)`. The commits that contain the changes are the following: [2bc5137](https://github.com/nasa/PROTEUS/pull/28/commits/2bc5137ac7c372c76a6da4bd48e409c5abd92481) and [5be73c2](https://github.com/nasa/PROTEUS/pull/28/commits/5be73c21ab2c47d001c9cd51d2545586a4264635) .

nemo794 replied:
Ah, that explains one of the comments from yesterday. (tbh, seeing the `, 1, 0)` was what caught my eye in the first place, because it felt like there could be a more Pythonic way.)

Glad we got to the bottom of it, and thank you for making these updates and the shift to `np.uint8`.

Looks good!

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -1246,7 +1365,8 @@ def _compute_diagnostic_tests(blue, green, red, nir, swir1, swir2,
 def _compute_mask_and_filter_interpreted_layer(
-        unmasked_interpreted_water_layer, qa_band):
+        unmasked_interpreted_water_layer, qa_band,
+        mask_adjacent_to_cloud_mode):
```
I think we want to provide all three options.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -2830,11 +3274,11 @@ def generate_dswx_layers(input_list,
+        invalid_ind = np.where(blue == image_dict['fill_data'])
     else:
-        invalid_ind = np.where(blue < -0.5)
+        invalid_ind = np.where(np.isnan(blue))
```
Yes, that's correct. Thanks for checking.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -1432,50 +1609,64 @@ def _load_hls_from_file(filename, image_dict, offset_dict, scale_dict,
         dswx_metadata_dict['SPACECRAFT_NAME'] = spacecraft_name
         if sensor is not None:
-            dswx_metadata_dict['SENSOR'] = sensor
+            # Sensor may be in the form: "OLI_TIRS; OLI_TIRS"
+            sensor_name = None
+            flag_all_same = True
+            sensor_list = []
+            for s in sensor.split(';'):
+                current_sensor_name = s.strip().replace('_TIRS', '')
+                sensor_list.append(current_sensor_name)
+                if sensor_name is None:
+                    sensor_name = current_sensor_name
+                    continue
+                if sensor_name != current_sensor_name:
+                    flag_all_same = False
+                    break
+            if flag_all_same:
+                dswx_metadata_dict['SENSOR'] = sensor_name
+            else:
+                dswx_metadata_dict['SENSOR'] = '; '.join(sensor_list)
```
If `SENSING_TIME` has more than one instance, it's expected that all instances will be different to each other. If `SENSOR` has more than one instance, it's expected that they are all the same. I think we can simplify this for the user. And no, I don't think we should force it to be `OLI` or `MSI`.

nemo794 replied:
Ah, ok. In that case, would it be possible to simplify this code section? Something like:

```
# Sensor may be in the form: "OLI_TIRS; OLI_TIRS"

# Tidy the names (DSWx-HLS does not use Landsat's TIR bands;
# only outputs from the OLI will be used in DSWx-HLS).
sensor_list = sensor.replace('_TIRS', '')

# split sensor name(s) apart, and use set() to remove redundancies.
sensor_list = set(sensor_names.split('; '))

dswx_metadata_dict['SENSOR'] = '; '.join(sensor_list)

```

This seems to produce the desired results for a singleton sensor, two sensors that are equal, and sensors that are not OLI nor MSI:

```
>>> sensor = "OLI_TIRS"
>>> sensor = sensor.replace('_TIRS', '')
>>> sensor = set(sensor.split('; '))
>>> '; '.join(sensor)
'OLI'

>>> sensor = "OLI_TIRS; OLI_TIRS"
>>> sensor = sensor.replace('_TIRS', '')
>>> sensor
'OLI; OLI'
>>> sensor = set(sensor.split('; '))
>>> sensor
{'OLI'}
>>> '; '.join(sensor)
'OLI'

>>> sensor = "OLI_TIRS; NEW_TIRS"
>>> sensor = sensor.replace('_TIRS', '')
>>> sensor = set(sensor.split('; '))
>>> sensor
{'OLI', 'NEW'}
>>> '; '.join(sensor)
'OLI; NEW'
```

gshiroma replied:
Hi @nemo794 , I like the suggestion! Thank you!

I just had one problem when running `set()`. It's not keeping the order:
```
set(['OLI', 'NEW'])
{'NEW', 'OLI'}
```
I'm not sure why my results where different from yours, maybe we are using different version of Python. Anyway, I fixed this by using a dictionary. I also changed a bit the code to handle spaces in `;`, e.g.,  `OLI_TIRS;OLI_TIRS`:
```
            sensor_names = sensor.replace('_TIRS', '')

            # Split sensor name(s) apart, and use set() to remove redundancies.
            sensor_list = [s.strip() for s in sensor_names.split(';')]
            sensor_list_unique = list(dict.fromkeys(sensor_list))
            dswx_metadata_dict['SENSOR'] = '; '.join(sensor_list_unique)
```

nemo794 replied:
Oh wow -- good catch that the order does not necessarily stay the same, I didn't realize -- and nice fix!

Should probably update the comment to:
```
# Split sensor name(s) apart, remove redundancies while maintaining order,
# and add to the metadata dict
```

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -174,12 +239,17 @@
                             "water heavy": [6, 3, 7, 1]}
 
-landcover_nir_threshold = 1200
+MIN_SLOPE_ANGLE = -5
+MAX_SUN_LOCAL_INC_ANGLE = 40
```
`MIN_SLOPE_ANGLE` and `MAX_SUN_LOCAL_INC_ANGLE` are default values for the `generate_dswx_layers()` function, which is not used by the DSWx-HLS SAS since the default values for the SAS come from the default runconfig.

I don't want to make `min_slope_angle` and `max_sun_local_inc_angle` required arguments for `generate_dswx_layers()`. Would you have any suggestions on how to improve this?

nemo794 replied:
Aye, the input handling makes this a tricky problem.

If we want the user to be able to change it via the runconfig, then perhaps something like was done for the browse image height / width?
https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358
It's not elegant, but it sets the defaults in a single location, and forces those values to propagate nicely via argument lists.

Alternatively, if these two parameters do not need to be exposed in the runconfig, then they could simply be removed from the .yaml's and from the input parsing, and left as the global constants.

Side note: While pondering this, I noticed that this has already caused a maintainability bug:
In the `_compute_opera_shadow_layer` function, `MIN_SLOPE_ANGLE` is used a default for the parameter `min_slope_angle`. But in the function body, only `MIN_SLOPE_ANGLE` (the global constant) is called, while the actual parameter `min_slope_angle` is never called. This means that any changes to the runconfig's `min_slope_angle` will be silently ignored during that function's execution.

gshiroma replied:
Hi @nemo794 , as far as I understood, you are doing the same here:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/defaults/dswx_hls.yaml#L94

BTW, maybe we should create constants for default browse images dimensions too:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358

Also, I'd prefer to keep the assignments above in the function call of `generate_dswx_layers()` in:

https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L2992

Would you mind helping me fix that? I think the bug you mentioned is already fixed, but if not, could you please fix that as well. Thanks!

nemo794 replied:
> Hi @nemo794 , as far as I understood, you are doing the same here:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/defaults/dswx_hls.yaml#L94
> 

Yes, a suggested value is included in the default runconfig.yaml file. However, as the code is currently written, if the user supplies any runconfig.yaml file, then the default runconfig.yaml values are completely ignored/overwritten during the call to `_deep_copy()`.

For example, say I'm a user who provides my own runconfig.yaml file, and fills in the input, output, etc. directories but leaves the parameter values blank. When the code is run, the `bin/dswx_hls.py` file will parse the arguments via `parse_runconfig_file()`. However, those values will not be filled in from the default runconfig.yaml file; instead they'll be parsed as (I think?) a Nonetype. (Will need to triple-check the type of the object -- I'm 95% sure it was Nonetype, but it's been a week since I ran these.)

The moral of this is that the .yaml file cannot be trusted to be the source of the defaults, since it gets ignored. We need to have the defaults be set within the `src/proteus/dswx_hls.py` code itself.

> BTW, maybe we should create constants for default browse images dimensions too:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358
> 

Creating global constants contradicts the goal of having only one source for the these values, from which those values propagate everywhere else in the code. Right now there are two sources: 1) the global constant, and 2) the runconfig.yaml file.

> Also, I'd prefer to keep the assignments above in the function call of generate_dswx_layers() in:
> https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L2992
> 

At least for the browse images, setting the default values here in the function signature is effectively meaningless when the function is called from the `bin/dswx_hls.py` script. It took a while to figure this out, but what's happening is that when `bin/dswx_hls.py` parses the arguments, the way the parsing function is set up, it will always make sure there is a value supplied for e.g. `args.browse_image_height` -- even if that value is `None`. Then, since `bin/dswx_hls.py` passes that known value of `args.browse_image_height` into `generate_dswx_layers()` for each parameter, that means that the default values supplied in the `generate_dswx_layers()` function signature will never actually be used by the current workflow. This was why I had to do to hacky workaround of [checking for "None"](https://github.com/gshiroma/PROTEUS/blob/036e33453547598f52371b32c8adf4336082f2c0/src/proteus/dswx_hls.py#L3355-L3358).

> Would you mind helping me fix that? I think the bug you mentioned is already fixed, but if not, could you please fix that as well. Thanks!

Ah, seems the version of the code I was looking at was a few hours out of date. You're right, the bug in `_compute_opera_shadow_layer` appears to be fixed now.

gshiroma replied:
Default values should come from the default runconfig. These values should be used by DSWx-HLS SAS if the user don't supply his own values, that's why it's a "default runconfig".

It would be great if `generate_dswx_layers()` could use the same values from the runconfig, but I'm not sure how to do that. Please, keep in mind that I'm trying to make `generate_dswx_layers()` callable from code different from the DSWx-HLS SAS so I don't want to make those parameters required. Since, I'm not sure how to do that, at least without an ugly workaround, I think we can use global constants. Notice that these global constants are never used by the DSWx-HLS SAS, as the default values for the SAS come from the default runconfig.

nemo794 replied:
100% agree that having all default values come from the default runconfig.yaml would be great, and `generate_dswx_layer()` should definitely have the ability to be called from anywhere -- those intentions have been very apparent from the beginning.

However, that's **not** the current behavior of the existing code and workflow, and the current code is the framework we have to work within. Fixing this properly would likely be a hefty undertaking.

I also understand that these global constants /should/ not be used by the DSWx-HLS SAS, but as mentioned in a previously, during the ongoing process of development, it's too easy for a developer (myself included) to accidentally use these global constants inside a function. When this happened in `_compute_opera_shadow_layer` a couple of days ago, you fortunately caught that bug, but it was caught by chance; relying on chance doesn't seem ideal.

As a quick workaround to set defaults within the existing framework, we could do a series of input checks at the beginning of the `generate_dswx_layers()` function, e.g.:

```
def generate_dswx_layers( ... ):
"""
...
"""
    # Set Defaults
    if min_slope_angle is None:
        min_slope_angle = -5
     
    if max_sun_local_inc_angle is None:
        min_slope_angle = 40

    if hls_thresholds is None:
        hls_thresholds = parse_runconfig_file()

    if scratch_dir is None:
        scratch_dir = '.'
     
     etc.
```

How about something like this?

gshiroma replied:
Uhm, the code that you suggested is setting these constants from two places as well (from the runconfig and from inside `generate_dswx_layers()`), so I'm not following the maintainability argument?

I also prefer to make these default values more explicit in the beginning of the module rather than buried inside the code.

nemo794 replied:
What I'm suggesting is not ideal. The ideal requires a significant updates to the input handling.

But, what I'm suggesting _does_ force the use of these values to be on a **single critical path** (runconfig first via `bin/dswx_hls.py`, which then must go through the input checks in `src/proteus/dsws_hls.py/generate_dswx_layers()`). Then, if a different function wants access to any one of these values, that function _must_ get the value from `src/proteus/dsws_hls.py/generate_dswx_layers()` by having it passed in via its argument list.

On the other hand, if the values are stored in a global constant, that global constant is **not** on that single critical path. This means that a different function could either get the value via its argument list (which is on the critical path), OR from the global constant. By having these two possible paths for a function to get the value, this is the source of the maintenance issues.

gshiroma replied:
They are still defined in two places and values can differ. I still prefer these constants explicit in the beginning of the module and set as default arguments in the function declaration. Much better than buried inside the code. When I find some time, I'll restructure the code so we can move on.

gshiroma replied:
This issue was addressed in [e7538c8](https://github.com/nasa/PROTEUS/pull/28/commits/e7538c8f1d936cd04ede6cf026bb51587aa2358d) .

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -648,63 +750,69 @@ def create_landcover_mask(copernicus_landcover_file,
     # Create urban-areas mask
-    logger.info(f'Creating urban-areas mask')
+    logger.info(f'creating urban-areas mask')
+    # WorldCover class 50: built-up
     urban_binary_mask = np.where((worldcover_array_up_3 == 50) , 1, 0)
```
Please, see comment above.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -648,63 +750,69 @@ def create_landcover_mask(copernicus_landcover_file,
     # Create vegetation mask
-    logger.info(f'Creating vegetation mask')
+    logger.info(f'creating vegetation mask')
+    # WorldCover class 10: tree cover
     tree_binary_mask  = np.where((worldcover_array_up_3 == 10) , 1, 0)
```
Please, see the comment above.

gshiroma commented:
> 

Thank you so much, @nemo794 , for doing this in-depth review in time for this cal/val point release. I think I addressed all your comments, please let me know if anything is missing.

gshiroma commented:
@hfattahi , if you have a chance, could you please take a look and see if you can approve this PR?

gshiroma commented:
@nemo794 , thank you again for your review. I think I addressed all your concerns. Please, let me know if anything is missing.

collinss-jpl reviewed (commented):
Hi @gshiroma, from my perspective the changes all look fine, but I would like to make a request for a small change to the DSWx metadata comparison that would help streamline our PGE integration testing:

The change I would like to see is at this line [here](https://github.com/gshiroma/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/dswx_hls.py#L670), where the metadata comparison routine checks for the `PROCESSING_DATETIME` key so it can be ignored, since this field will always be different between runs. 

I would like to see check expanded to include the `DEM_FILE`, `WORLDCOVER_FILE` and `LANDCOVER_FILE` keys since these fields are only file paths and often need to be different paths in our testing vs what was used in the expected results. For example, we currently use the dswx_compare.py script with our integration test Jenkins pipeline for DSWx, but we always "fail" the comparison since the ancillary files are kept in different locations from the expected data. All the actual image array checks pass though.

I propose that the condition becomes: `if k1 in ('PROCESSING_DATETIME', 'DEM_FILE', 'WORLDCOVER_FILE', 'LANDCOVER_FILE'): continue`

Hopefully this is not too big a change to include in this PR, or if it cannot make it into this release, I hope it can be included eventually.

gshiroma commented:
> Hi @gshiroma, from my perspective the changes all look fine, but I would like to make a request for a small change to the DSWx metadata comparison that would help streamline our PGE integration testing:
> 
> The change I would like to see is at this line [here](https://github.com/gshiroma/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/dswx_hls.py#L670), where the metadata comparison routine checks for the `PROCESSING_DATETIME` key so it can be ignored, since this field will always be different between runs.
> 
> I would like to see check expanded to include the `DEM_FILE`, `WORLDCOVER_FILE` and `LANDCOVER_FILE` keys since these fields are only file paths and often need to be different paths in our testing vs what was used in the expected results. For example, we currently use the dswx_compare.py script with our integration test Jenkins pipeline for DSWx, but we always "fail" the comparison since the ancillary files are kept in different locations from the expected data. All the actual image array checks pass though.
> 
> I propose that the condition becomes: `if k1 in ('PROCESSING_DATETIME', 'DEM_FILE', 'WORLDCOVER_FILE', 'LANDCOVER_FILE'): continue`
> 
> Hopefully this is not too big a change to include in this PR, or if it cannot make it into this release, I hope it can be included eventually.

Hi Scott, this is great feedback and suggestions. We renamed 'DEM_FILE', 'WORLDCOVER_FILE', and 'LANDCOVER_FILE' as 'DEM_SOURCE, 'WORLDCOVER_SOURCE',  and 'LANDCOVER_SOURCE', respectively. The idea is that the filename probably doesn't provide good description about the contents of the files. So for example, instead of using the file name for the DEM `EPSG4326.vrt`, we'll pass to the DSWx-HLS SAS a string containing a description such as "Copernicus DEM GLO-30 2021" or something similar. The fields with suffix `_SOURCE` will come preferably from the fields in the runconfig named with suffix `_DESCRIPTION`. Only if those fields are empty that we'll use the fields with suffix `_FILE` to populate the metadata.

See the updated runconfig [here](https://github.com/nasa/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/defaults/dswx_hls.yaml)

Considering that, would you like that we add 'DEM_SOURCE, 'WORLDCOVER_SOURCE',  and 'LANDCOVER_SOURCE' to the list of exceptions along with 'PROCESSING_DATETIME'?

hfattahi commented on src/proteus/dswx_hls.py:
```diff
@@ -11,15 +11,35 @@
 from proteus.core import save_as_cog
 
-PRODUCT_VERSION = '0.1'
+PRODUCT_VERSION = '0.5'
```
This is not really the product version. It looks like you mean the PROTEOUS sofwtare version. It should not be called "PRODUCT_VERSION" as this should be really defined by a collection of representatives from Algorithm, validation and data system teams. 
I think we agreed that the product version should be an input to the run config file which SDS can populate using an external knowledge.
The software version should be easily derived from our software "e.g., proteou.__version__" or something similar.
Ideally if PCM and PGE could pass the PCM and PGE versions to the SAS through the run config, then we could store all the version infomration we need to track of. To summarize if I'm not mistaken we decided to track multiple versions:

1- product version
2- SAS version 
3- PGE version
4- PCM version
5- ancillary data version

@collinss-jpl @cmarshak can you guys chime in on what is our final conclusion on versioning

cmarshak replied:
I agree with what you said Heresh. I don't know the precise nomenclature that you and Scott and others have agreed on.

I do think `PRODUCT_VERSION` should be from the point of the end-user and not be related to the software so should remain at 0.1 until production begins. Just make sure the other version categories (in particular, the proteus software version) are tracked and are saved downstream in the ES database.

Currently, redeployment in SDS wipes all the previously generated datasets. In the future, this will not be the case so being able to facet on the versions will be more important. In particular, being able to differentiate datasets generated by different versions of PROTEUS, in whatever name that is appropriate.

gshiroma replied:
@hfattahi, @collinss-jpl , and @cmarshak , thank you for your inputs. I just pushed commit [39cb5ae](https://github.com/nasa/PROTEUS/pull/28/commits/39cb5ae2d742f04022990e604d663c9eeaec1b47) that addresses your comments. I exposed the product version in the runconfig. This field will be used to populate the `PRODUCT_VERSION` in the DSWx-HLS metadata. I also added `SOFTWARE_VERSION` to the DSWx-HLS metadata, which is populated by the PROTEUS version (`v0.5` in this case). Please let me know if this is what you expect and if you have any further comments or concerns.

hfattahi replied:
Thanks @gshiroma That is great. @collinss-jpl note that product_version should be populated by PCM. 

The Software version needs to be automated in the final release. I see it hard coded for now which is fine. 

The only part that may still be missing is the SDS version which I think should represent PCM, PGE and ideally the ancillary data. Perhaps that can be added in the final release.

hfattahi reviewed (commented):
Thanks @gshiroma . It looks good to me as far as I can tell. I put in a comment about product version and versioning in general since we discussed that in details.

collinss-jpl commented:
> > Hi @gshiroma, from my perspective the changes all look fine, but I would like to make a request for a small change to the DSWx metadata comparison that would help streamline our PGE integration testing:
> > The change I would like to see is at this line [here](https://github.com/gshiroma/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/dswx_hls.py#L670), where the metadata comparison routine checks for the `PROCESSING_DATETIME` key so it can be ignored, since this field will always be different between runs.
> > I would like to see check expanded to include the `DEM_FILE`, `WORLDCOVER_FILE` and `LANDCOVER_FILE` keys since these fields are only file paths and often need to be different paths in our testing vs what was used in the expected results. For example, we currently use the dswx_compare.py script with our integration test Jenkins pipeline for DSWx, but we always "fail" the comparison since the ancillary files are kept in different locations from the expected data. All the actual image array checks pass though.
> > I propose that the condition becomes: `if k1 in ('PROCESSING_DATETIME', 'DEM_FILE', 'WORLDCOVER_FILE', 'LANDCOVER_FILE'): continue`
> > Hopefully this is not too big a change to include in this PR, or if it cannot make it into this release, I hope it can be included eventually.
> 
> Hi @collinss-jpl , this is great feedback and suggestions. We renamed 'DEM_FILE', 'WORLDCOVER_FILE', and 'LANDCOVER_FILE' as 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE', respectively. The idea is that the filename probably doesn't provide good description about the contents of the files. So for example, instead of using the file name for the DEM `EPSG4326.vrt`, we'll pass to the DSWx-HLS SAS a string containing a description such as "Copernicus DEM GLO-30 2021" or something similar. The fields with suffix `_SOURCE` will come preferably from the fields in the runconfig named with suffix `_DESCRIPTION`. Only if those fields are empty that we'll use the fields with suffix `_FILE` to populate the metadata.
> 
> See the updated runconfig [here](https://github.com/nasa/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/defaults/dswx_hls.yaml)
> 
> Considering that, would you like that we add 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE' to the list of exceptions along with 'PROCESSING_DATETIME'?

If 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE' will be populated based on fields set in the RunConfig, then we should be able to ensure we can match the expected results during the metadata comparison. If DEM_FILE, WORLDCOVER_FILE, LANDCOVER_FILE are being kept as backups, they should still get ignored by the metadata comparison.

gshiroma commented:
> > > Hi @gshiroma, from my perspective the changes all look fine, but I would like to make a request for a small change to the DSWx metadata comparison that would help streamline our PGE integration testing:
> > > The change I would like to see is at this line [here](https://github.com/gshiroma/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/dswx_hls.py#L670), where the metadata comparison routine checks for the `PROCESSING_DATETIME` key so it can be ignored, since this field will always be different between runs.
> > > I would like to see check expanded to include the `DEM_FILE`, `WORLDCOVER_FILE` and `LANDCOVER_FILE` keys since these fields are only file paths and often need to be different paths in our testing vs what was used in the expected results. For example, we currently use the dswx_compare.py script with our integration test Jenkins pipeline for DSWx, but we always "fail" the comparison since the ancillary files are kept in different locations from the expected data. All the actual image array checks pass though.
> > > I propose that the condition becomes: `if k1 in ('PROCESSING_DATETIME', 'DEM_FILE', 'WORLDCOVER_FILE', 'LANDCOVER_FILE'): continue`
> > > Hopefully this is not too big a change to include in this PR, or if it cannot make it into this release, I hope it can be included eventually.
> > 
> > 
> > Hi @collinss-jpl , this is great feedback and suggestions. We renamed 'DEM_FILE', 'WORLDCOVER_FILE', and 'LANDCOVER_FILE' as 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE', respectively. The idea is that the filename probably doesn't provide good description about the contents of the files. So for example, instead of using the file name for the DEM `EPSG4326.vrt`, we'll pass to the DSWx-HLS SAS a string containing a description such as "Copernicus DEM GLO-30 2021" or something similar. The fields with suffix `_SOURCE` will come preferably from the fields in the runconfig named with suffix `_DESCRIPTION`. Only if those fields are empty that we'll use the fields with suffix `_FILE` to populate the metadata.
> > See the updated runconfig [here](https://github.com/nasa/PROTEUS/blob/d31e0212a0d4d680826059d66ae3177238bc6fb3/src/proteus/defaults/dswx_hls.yaml)
> > Considering that, would you like that we add 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE' to the list of exceptions along with 'PROCESSING_DATETIME'?
> 
> If 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE' will be populated based on fields set in the RunConfig, then we should be able to ensure we can match the expected results during the metadata comparison. If DEM_FILE, WORLDCOVER_FILE, LANDCOVER_FILE are being kept as backups, they should still get ignored by the metadata comparison.

Hi @collinss-jpl , I just added 'DEM_SOURCE, 'WORLDCOVER_SOURCE', and 'LANDCOVER_SOURCE' to the list of exceptions along with 'PROCESSING_DATETIME' (see [000df34](https://github.com/nasa/PROTEUS/pull/28/commits/000df34b70cc3495e0f70060157aaca528ca6500) ). The fields `DEM_FILE`, `WORLDCOVER_FILE`, and `LANDCOVER_FILE` were removed from the DSWx-HLS product metadata.

gshiroma commented:
@nemo794 , excellent catch regarding `_deep_update()`. I just fixed it in commit [47dff51](https://github.com/nasa/PROTEUS/pull/28/commits/47dff512867495e2c68d98fa85b2fc8171ffaef8) to ignore fields with `None` from the user runconfig and read them from the default runconfig. Thank you for catching that!

hfattahi reviewed (approved):
Thank you @gshiroma . Also thanks to @nemo794 for adding the browse image part.

gshiroma commented:
> Thank you @gshiroma . Also thanks to @nemo794 for adding the browse image part.

Thank you, @hfattahi , for your review and comments!

gshiroma commented:
@nemo794 , I checked, and there seems to be no issue with getting RunConfigConstants() from argparse arguments. See example below:
```
>> dswx_hls.py runconfig.yaml --bheight--browse-image-height 100 --bwidth--browse-image-width 200 --min-slope-angle -10 --max-sun-local-inc-angle 50
Default runconfig file: <removed>/proteus/defaults/dswx_hls.yaml
YAML schema: <removed>/proteus/schemas/dswx_hls.yaml
Input runconfig file: l30_greenland_west_2/dswx_hls_adj_ignore.yaml
Validating runconfig file: l30_greenland_west_2/dswx_hls_adj_ignore.yaml
HLS thresholds:
     wigt: 0.124
     awgt: 0.0
     pswt_1_mndwi: -0.44
     pswt_1_nir: 1500
     pswt_1_swir1: 900
     pswt_1_ndvi: 0.7
     pswt_2_mndwi: -0.5
     pswt_2_blue: 1000
     pswt_2_nir: 2500
     pswt_2_swir1: 3000
     pswt_2_swir2: 1000
     lcmask_nir: 1200
input HLS files directory: l30_greenland_west_2/
input parameters:
    file(s):
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.VAA.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B09.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.SAA.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.SZA.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B03.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B02.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B01.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.VZA.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B05.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B11.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B10.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B04.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.Fmask.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B06.tif
        l30_greenland_west_2/HLS.L30.T22VEQ.2021248T143156.v2.0.B07.tif
    DEM file: <removed>/nisar-dem-copernicus-2/EPSG4326/EPSG4326.vrt
    scratch directory: scratch_dir
product parameters:
    product ID: dswx_hls
    product version: 0.5
    software version: 0.5
processing parameters:
    flag_use_otsu_terrain_masking: False
    min_slope_angle: -10.0
    max_sun_local_inc_angle: 50.0
    mask_adjacent_to_cloud_mode: ignore
browse image:
    browse_image_height: 100
    browse_image_width: 200
```
Please, let me know if there's anything to get this PR merged.

collinss-jpl reviewed (approved):
Thanks @gshiroma for updating the metadata comparison! @hfattahi, regarding product version population, we should have that implemented on the PCM soon.

nemo794 commented:
> @nemo794 , I checked, and there seems to be no issue with getting RunConfigConstants() from argparse arguments. See example below:
> 
> (see above)
> 
> Please, let me know if there's anything missing to get this PR merged.

@gshiroma What is being demonstrated here is not actually my concern. The above demo is illustrates what values eventually make it to these lines of code: ([link](https://github.com/gshiroma/PROTEUS/blob/99ca1f96bc514f50a81b7714c74fb4f75fbb9192/src/proteus/dswx_hls.py#L3179-L3199)). It does not demonstrate fully where those values are coming from, nor how the code is making decisions about which values from the three potential input sources is being used.

My concern is two fold: First: simplifying, streamlining, and creating a consistency between the fields of where the user supplies the input value. Second: internal to the code, it'd be great to lean into coding practices are more robust and maintainable, in order to prevent accidental - but preventable! - bugs in the future.

However, that seems to be a bigger discussion than there is time for in this point release. I'll open an issue.

nemo794 reviewed (approved):
The proof is in the pudding (so to speak) that there was a huge amount of effort that went into getting DSWx-HLS ready for this point release. Many new features were created and implemented, several bugs were caught, and the code comments and clarity improved. Really, really nice! Thank you!

Offline, we've already started discussions about ongoing improvements for future releases. But for this release, things are working well. Onwards and upwards!

gshiroma commented:
> The proof is in the pudding (so to speak) that there was a huge amount of effort that went into getting DSWx-HLS ready for this point release. Many new features were created and implemented, several bugs were caught, and the code comments and clarity improved. Really, really nice! Thank you!
> 
> Offline, we've already started discussions about ongoing improvements for future releases. But for this release, things are working well. Onwards and upwards!

Thank you, @nemo794 . Yes, I'll definitely need more help from you to improve this software. Thank you so much for you review and please create issues to address your concerns or submit pull requests to fix them directly.
